rpc: add rpc to get mempool txs spending specific prevouts

t-bast opened the pull request:
We add an RPC to fetch mempool transactions spending any of the given outpoints.

Without this RPC, application developers need to first call `getrawmempool` which returns a long list of `txid`, then fetch each of these transactions individually (`getrawtransaction`) to check whether they spend the given outpoints, which wastes a lot of bandwidth (in the worst case we need to transfer the whole mempool).

For example in lightning, when we discover that one of our channel funding transactions has been spent, we need to find the spending transaction to claim our outputs from it. We are currently forced to fetch the whole mempool to do the analysis ourselves, which is quite costly.

I believe that this RPC is also generally useful when doing some introspection on your mempool after one of your transactions failed to broadcast, for example when you implement RBF at the application level. Fetching and analyzing the conflicting transaction gives you more information to successfully replace it.

t-bast commented on test/functional/mempool_packages.py:
```diff
@@ -100,6 +100,11 @@ def run_test(self):
             entry = self.nodes[0].getmempoolentry(x)
             assert_equal(entry, mempool[x])
 
+            # Check that getmempoolspendingtx is consistent with getrawmempool
```
It looks like there aren't any dedicated test files for `getmempoolentry`, `getmempoolancestors` and `getmempooldescendants` (I wanted to insert my tests in such a file).

Should I create a new `rpc_getmempoolpendingtx.py` to put more tests (e.g. some failure cases)?

glozow replied:
Yeah, that's probably the move. "rpc_mempool_info.py" would be a good place to put tests for all of these RPCs.

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/24408/commits/4ef5a16c306bb4784fb645d1df51fe0efeadcb06, I have only added tests for my new RPC for now, future PRs can add tests for `getmempoolentry`, `getmempoolancestors` and `getmempooldescendants` here if that makes sense.

darosior commented:
For this usecase, why not using `getmempoolentry`'s `spentby` field?

t-bast commented:
> For this usecase, why not using getmempoolentry's spentby field?

Because that wouldn't work for confirmed utxos!

My funding transaction (or any L2 protocol initial state) has been confirmed on-chain and I want to check who spends it in the mempool. I can't use `getmempoolentry` because I don't know yet the `txid` of the spending tx...

darosior commented:
Right, sorry, i shouldn't have inferred the behaviour from the name. I see it's in the context of `txindex` then, so i don't have an opinion.

t-bast commented:
It's much weaker than `txIndex`, it's exposing a `txIndex` for mempool contents, which is why I think it's acceptable (I'm only querying existing mempool functions).

prayank23 commented:
> > For this usecase, why not using getmempoolentry's spentby field?
> 
> Because that wouldn't work for confirmed utxos!
> 
> My funding transaction (or any L2 protocol initial state) has been confirmed on-chain and I want to check who spends it in the mempool. I can't use `getmempoolentry` because I don't know yet the `txid` of the spending tx...
> 
> Maybe the RPC name is misleading, because its inputs (`txid` and `vout`) aren't necessarily for a mempool tx (and most of the time will not), but the RPC output will be a mempool tx (or `null`).

Is it possible to add it in existing RPC `getmempoolenfry` by adding more arguments (optional)?

t-bast commented:
> Is it possible to add it in existing RPC getmempoolentry by adding more arguments (optional)?

I'm not sure it would be very ergonomic, because `getmempoolentry` requires the `txid` of the mempool transaction you're fetching. If we wanted to overload it, it would have the following two distinct sets of behaviors:

- a `txid` is provided, so we return a mempool transaction that matches this `txid`
- a `txid` (probably named differently to avoid confusion) and a `vout` are provided, so we return a mempool transaction spending this given `txid:vout`

Correctly defining the arguments lists (and which are required) to support both use-cases without creating confusion seems hard to be honest...

nopara73 commented:
Not being able to walk the chain forward is certainly the source of huge friction and hacky code while working with Bitcoin RPC. I'd even go that far to claim that this is the single largest missing feature that if we'd have, it would have saved the most time for me personally. Certainly cACK this approach, but I'd like to note a general version of it would be more useful, somehow we should be able to determine which transaction spends a UTXO if any, regardless if it's from mempool or not.

glozow commented on src/rpc/blockchain.cpp:
```diff
@@ -786,6 +786,49 @@ static RPCHelpMan getmempoolentry()
+static RPCHelpMan getmempoolspendingtx()
+{
+    return RPCHelpMan{"getmempoolspendingtx",
+                "\nScans the mempool to find a transaction spending the given output\n",
+                {
+                    {"txid", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "transaction id"},
+                    {"n", RPCArg::Type::NUM, RPCArg::Optional::NO, "vout number"},
+                },
```
I'm wondering if, in the future, you might want to pass in a list of prevouts? (e.g. if you have transaction(s) with multiple inputs and want to see if anything in the mempool conflicts)

t-bast replied:
That's a good idea, it's worth adding this flexibility right now to avoid running in backwards-compatibility issues if we add it later, I'll try that. I'll take a list of `prevouts` and return a map of `prevout -> spending tx (or null if not found)`.

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/24408/commits/4ef5a16c306bb4784fb645d1df51fe0efeadcb06
I chose to return only the `spendingtxid` instead of the mempool tx data, because otherwise if a transaction appears multiple times (because it spends many of the provided prevouts) it would be duplicated, which is a (small) waste.

Callers can chain this with `getmempoolentry` to get more data about the conflicting transaction.
If you think it makes more sense to directly include the mempool tx let me know, I don't have strong feelings about that.

glozow commented on src/rpc/blockchain.cpp:
```diff
@@ -786,6 +786,49 @@ static RPCHelpMan getmempoolentry()
     };
 }
 
+static RPCHelpMan getmempoolspendingtx()
```
Sorry to nitpick but I was confused in the beginning by the name as well (I thought you wanted the child of a mempool transaction). Maybe `gettxspendingprevout` would be better, but I'm sure there's someone out there who would find it even more confusing :sweat_smile:

t-bast replied:
I agree, I'm not a fan of the name! Should it be `getmempooltxspendingprevout` though to highlight that we return a mempool tx (with ancestor / descendant feerates, which are useful in fee-bumping cases)? It's a long name, but at least it's quite obvious what it does (hopefully).

t-bast replied:
I updated the name, but now that is accepts multiple prevouts, I'm not sure whether I should make it plural or not...
None of the names I find are completely satisfying, so I'll let reviewers chime in with better proposals!

glozow reviewed (commented):
Concept ACK, seems very reasonable to me especially since we already have a mempool function to fetch the transaction. Also in general, it seems useful to be able to fetch the transaction that conflicts with yours.

glozow commented:
> I'd like to note a general version of it would be more useful, somehow we should be able to determine which transaction spends a UTXO if any, regardless if it's from mempool or not.

This might be tricky (the non-mempool case) since we would have deleted the UTXO from our chainstate. And not possible sometimes on a pruned node.

michaelfolkson commented:
Concept ACK - rationale in PR description seems strong and other projects (e.g. Wasabi) have experienced same issue.

> I'd like to note a general version of it would be more useful, somehow we should be able to determine which transaction spends a UTXO if any, regardless if it's from mempool or not

I think the use case from @nopara73's comment above would be you don't know whether the transaction spending the UTXO is in your mempool or confirmed onchain and you want to query your mempool first and then your transaction index? I think generally you'd want one RPC to query your mempool and one RPC to query your transaction index. Combining both into a single RPC seems to me against separation of concerns (and is easily resolved by just running two RPC commands).

nopara73 commented:
FTR there's `gettxout`. I think my suggestion would fit best there.

t-bast commented:
> FTR there's gettxout. I think my suggestion would fit best there.

@nopara73 do you mean that instead of adding a new RPC like I'm doing, we could update `gettxout` to return the spending tx when the utxo is spent?

In the case where the spending tx is not in the mempool, I think the only solution would be to walk the chain backwards and analyze each block to see if the spending tx is inside it. However, as @glozow says, if you're running a pruned node, you will likely analyze every block you have and won't find anything, which is very inefficient. Even worse, if the provided utxo never existed, you will walk the whole chain without finding anything...

It looks like a much more debatable change which would involve serious trade-offs, that's why I think restricting this PR to mempool txs has a better chance of being accepted. I could still reuse `gettxout` instead of adding a new RPC, but that would make it harder (backwards-compatibility wise) to accept a list of `prevouts` as @glozow suggests [here](https://github.com/bitcoin/bitcoin/pull/24408#discussion_r811843082).

w0xlt reviewed (commented):
Concept ACK.
Maybe the two commits can be merged. Apparently the first commit is in an invalid state (since the `outputs` parameter is not converted).

t-bast commented:
> Maybe the two commits can be merged.

Definitely, done. I was waiting for the builds to complete to check that I hadn't missed something else that would need fixing.

MarcoFalke commented:
> Without this RPC, application developers need to first call getrawmempool which returns a long list of txid, then fetch each of these transactions individually (getrawtransaction) to check whether they spend the given outpoints, which wastes a lot of bandwidth (in the worst case we need to transfer the whole mempool).

I think without this RPC, you can create a tx that spends the input and only fetch the whole mempool if the input was spent.

Moreover, I wonder what happens if the tx is included in a block right away? So the input will be gone, but there isn't a tx in the mempool that spends it.

While the RPC makes sense on its own, it doesn't fix your problem completely?

MarcoFalke commented on doc/release-notes.md:
```diff
@@ -123,6 +123,8 @@ New RPCs
   now reflects the status of the current block rather than the next
   block. (#23508)
+- A new `getmempooltxspendingprevout` RPC has been added, which scans the mempool to find
+  transactions spending any of the given outpoints. (#24408)
```
Maybe move this to a separate file for now?

`doc/release-notes-24408.md`

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/24408/commits/4c96203ec36f5634d6f73f3b9a1d99bbd91dcc21, let me know if the format should be different

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -786,6 +786,90 @@ static RPCHelpMan getmempoolentry()
+static RPCHelpMan getmempooltxspendingprevout()
+{
+    return RPCHelpMan{"getmempooltxspendingprevout",
+                "\nScans the mempool to find transactions spending any of the given outputs\n",
```
```suggestion
                "Scans the mempool to find transactions spending any of the given outputs",
```

nit: I think those can be removed

t-bast replied:
Agreed, done in https://github.com/bitcoin/bitcoin/pull/24408/commits/4c96203ec36f5634d6f73f3b9a1d99bbd91dcc21

t-bast commented:
Thanks for the review @MarcoFalke!

> I think without this RPC, you can create a tx that spends the input and only fetch the whole mempool if the input was spent.
> Moreover, I wonder what happens if the tx is included in a block right away? So the input will be gone, but there isn't a tx in the mempool that spends it.
> While the RPC makes sense on its own, it doesn't fix your problem completely?

I probably tried to be a bit too general in my description, let me try to explain in more details my current scenario.

In the specific case where I want to apply this in the short term, I'm trying to spend lightning htlcs. I have a pre-signed transaction that spends the outpoint I'm interested in but doesn't provide enough fees, so I'll want to add more inputs to reach the right feerate.

The complete flow looks like this:

- I verify that the output is unspent with `gettxout` (without including the mempool)
  - If it's spent, then I have nothing to do (actually it triggers another flow to spend the output of that spending tx, but that's orthogonal)
  - Otherwise, I check if there is already a mempool tx spending this output with `getmempooltxspendingprevout`
    - If nothing is in the mempool, I'll use `fundrawtransaction` to add inputs and broadcast the transaction
    - Otherwise, I'll modify the mempool transaction to pay more fees (by lowering/removing the existing change output and potentially adding new inputs)

Without this RPC, I would have to optimistically fund and broadcast, and if that fails fetch the whole mempool to find the conflicting transaction. I expect to regularly have to go through the RBF flow, so it feels inefficient to have to dump the mempool when that happens :(

You're right that if the current mempool transaction confirms while I'm funding its replacement, I'll need to correctly handle it, but that doesn't seem to be too much of a problem for my use-case.

Does that make sense with this more detailed scenario?

t-bast commented:
Actually there is a simpler place where we'll use that new RPC in our codebase. When a lightning node restarts, it lists its channels from its internal DB and verifies that the funding outpoint of each channel is still unspent (with `gettxout`).

If we discover than a channel has actually been spent (maybe our counterparty is malicious and saw that we were offline, and tried to cheat), we want to find the spending transaction to be able to react and publish our set of follow-up transactions. So the first thing we do is look into the mempool, currently by dumping it and checking every transaction ourselves. With `getmempooltxspendingprevout` we would avoid a performance hit here.

Of course, if the spending transaction isn't in the mempool, then we have to check recent blocks one by one, but that shouldn't happen often, and since lightning nodes shouldn't be offline for too long, you usually only have to check 1 or 2 blocks (which means transferring < 5MB which is much smaller than the whole mempool).

prayank23 commented on doc/release-notes/release-notes-24408.md:
```diff
@@ -0,0 +1,5 @@
+New RPCs
+--------
+
+- A new `getmempooltxspendingprevout` RPC has been added, which scans the mempool to find
```
```suggestion
- A new `getmempoolprevout` RPC has been added, which scans the mempool to find
```

danielabrozzoni reviewed (approved):
tACK 4c96203ec36f5634d6f73f3b9a1d99bbd91dcc21 - the code looks good to me, I tested locally and it works as expected.

I would have needed this API quite a few times in the past, so I'd be happy to see this merged :)

MarcoFalke commented:
> This seems like it should be using a watch-only wallet instead.

This wouldn't work for users that don't have or don't want a bdb/sqlite dependency as well the overhead of having to deal with wallet maintenance. So I can see the argument that this RPC makes it easier for users.

t-bast commented on src/rpc/blockchain.cpp:
```diff
@@ -786,6 +786,90 @@ static RPCHelpMan getmempoolentry()
+                        {
+                            {RPCResult::Type::STR_HEX, "txid", "the transaction id of the spent output"},
+                            {RPCResult::Type::NUM, "vout", "the vout value of the spent output"},
+                            {RPCResult::Type::STR_HEX, "spendingtxid", "the transaction id of the mempool transaction spending this output (ommitted if unspent)"},
```
Thanks! Fixed in https://github.com/bitcoin/bitcoin/pull/24408/commits/9a9011c5aa4a9d069ba897b07ff9792349692cbb

t-bast commented:
> This seems like it should be using a watch-only wallet instead.

@luke-jr can you detail how that would work? I may be missing something, but I don't see how that would address my issue. I do need a regular wallet as I'm relying heavily on `fundrawtransaction`, it would be wasteful to need both a watch-only wallet and a regular wallet to make this use-case work?

t-bast commented:
Sorry for the delay in responding to comments and rebasing, I've been away for a while, the PR is now rebased.

One of the points that I'd like more feedback on is that I chose to return only the `spendingtxid` instead of the mempool tx data, because otherwise if a transaction appears multiple times (because it spends many of the provided prevouts) it would be duplicated, which is a (small) waste.

Callers can chain this with `getmempoolentry` to get more complete tx data for the conflicting transaction. If you think it makes more sense to directly include the mempool tx let me know, I don't have strong feelings about that.

vincenzopalazzo commented on src/rpc/mempool.cpp:
```diff
@@ -376,6 +376,90 @@ RPCHelpMan getmempoolentry()
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, vout cannot be negative");
+        }
+
+        const COutPoint output(txid, nOutput);
+        outputs.push_back(output);
```
```suggestion
        outputs.emplace_back(txid, nOutput);
```

Change this can avoid a copy of the object

t-bast replied:
Thanks, done in https://github.com/bitcoin/bitcoin/pull/24408/commits/8e6db71ff6cf2c8f724183996b4842fcf53d16b0

t-bast replied:
The build started failing after this change, so I'm reverting it.
To be honest, I don't think we care about such optimizations on just a few bytes here.

vincenzopalazzo replied:
>The tests started failing after this change, so I'm reverting it.

Mh! strange the Ci looks like failing again.

>To be honest, I don't think we care about such optimizations on just a few bytes here.

Right, this can be also a followup PR too

MarcoFalke replied:
Test failure is due to stricter documentation checking in current master, not due to commit 8e6db71ff6cf2c8f724183996b4842fcf53d16b0 . I left a comment with a potential fix.

vincenzopalazzo commented on src/rpc/mempool.cpp:
```diff
@@ -376,6 +376,90 @@ RPCHelpMan getmempoolentry()
+
+    UniValue result{UniValue::VARR};
+
+    for (const COutPoint& output : outputs) {
+        UniValue o(UniValue::VOBJ);
+        o.pushKV("txid", output.hash.ToString());
+        o.pushKV("vout", (uint64_t)output.n);
+
+        const CTransaction* spendingTx = mempool.GetConflictTx(output);
+        if (spendingTx != nullptr) {
+            o.pushKV("spendingtxid", spendingTx->GetHash().ToString());
+        }
+
+        result.push_back(o);
+    }
```
```suggestion
    for (const COutPoint& output : outputs) {
        result.emplace_back(UniValue::VOBJ);
        UniValue o = result.back();
        o.pushKV("txid", output.hash.ToString());
        o.pushKV("vout", (uint64_t)output.n);

        const CTransaction* spendingTx = mempool.GetConflictTx(output);
        if (spendingTx != nullptr) {
            o.pushKV("spendingtxid", spendingTx->GetHash().ToString());
        }
    }
```

In this way we avoid the move too, and for a transaction this is important IMHO

t-bast replied:
Thanks for the review! This doesn't seem to compile though...and I'm not sure it's needed at all, we're not dealing with a whole transaction here, just a `txid`, so a copy is really not something we care about IMO.

vincenzopalazzo replied:
>Thanks for the review! This doesn't seem to compile though

I will try to compile it locally maybe something wrong with `UniValue::VOBJ` that my mind compiler is not able to catch right now!

>just a txid, so a copy is really not something we care about IMO

Right, I wrote a bad motivation, I mean here there `COutPoint` can be a lot, and this loop with the `emplace_back` and be optimized, because with the `push_back` to need to copy the structure in a pointer this can cause move, or a copy in some case.

With emplace_back you need to allocate the space of the struct one time, and this will summarize from:

push_back: 1 stack allocation + 1 copy to heap, and the copy cost O(N)
emplace_back: 1 allocation and this with reserve() cost O(1)

But yes, overall it is not a big deal.

vincenzopalazzo reviewed (changes requested):
Concept ACK

Just some code styling improvement

t-bast commented on src/rpc/mempool.cpp:
```diff
@@ -587,6 +587,90 @@ static RPCHelpMan getmempoolentry()
+                        {
+                            {RPCResult::Type::STR_HEX, "txid", "the transaction id of the spent output"},
+                            {RPCResult::Type::NUM, "vout", "the vout value of the spent output"},
+                            {RPCResult::Type::STR_HEX, "spendingtxid", "the transaction id of the mempool transaction spending this output (omitted if unspent)"},
```
Thanks for the tip, this seems to be working! Love the new github name btw ;)

vincenzopalazzo reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24408/commits/e143e80fb93a752f522c7cc3e98a4f55626d6b49

MarcoFalke commented on test/functional/rpc_mempool_info.py:
```diff
@@ -0,0 +1,98 @@
+        self.setup_clean_chain = True
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
```
I've reworked this test to use the MiniWallet.

If you want you can take the diff and squash. It requires a rebase on current master, though.

However, for some reason one line in the test fails and I fail to see why. I can take another look there, unless someone beats me to it.

Also, there are some whitespace fixes. You can check them with `git diff --ignore-all-space`.

```diff
diff --git a/src/rpc/mempool.cpp b/src/rpc/mempool.cpp
index fe25dc9edc..2d07b8bd2e 100644
--- a/src/rpc/mempool.cpp
+++ b/src/rpc/mempool.cpp
@@ -590,83 +590,83 @@ static RPCHelpMan getmempoolentry()
 static RPCHelpMan getmempooltxspendingprevout()
 {
     return RPCHelpMan{"getmempooltxspendingprevout",
-                "Scans the mempool to find transactions spending any of the given outputs",
+        "Scans the mempool to find transactions spending any of the given outputs",
+        {
+            {"outputs", RPCArg::Type::ARR, RPCArg::Optional::NO, "The transaction outputs that we want to check, and within each, the txid (string) vout (numeric).",
                 {
-                    {"outputs", RPCArg::Type::ARR, RPCArg::Optional::NO, "The transaction outputs that we want to check, and within each, the txid (string) vout (numeric).",
+                    {"", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED, "",
                         {
-                            {"", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED, "",
-                                {
-                                    {"txid", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "The transaction id"},
-                                    {"vout", RPCArg::Type::NUM, RPCArg::Optional::NO, "The output number"},
-                                },
-                            },
+                            {"txid", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "The transaction id"},
+                            {"vout", RPCArg::Type::NUM, RPCArg::Optional::NO, "The output number"},
                         },
                     },
                 },
-                RPCResult{
-                    RPCResult::Type::ARR, "", "",
-                    {
-                        {RPCResult::Type::OBJ, "", "",
-                        {
-                            {RPCResult::Type::STR_HEX, "txid", "the transaction id of the spent output"},
-                            {RPCResult::Type::NUM, "vout", "the vout value of the spent output"},
-                            {RPCResult::Type::STR_HEX, "spendingtxid", /*optional=*/true, "the transaction id of the mempool transaction spending this output (omitted if unspent)"},
-                        }},
-                    }
-                },
-                RPCExamples{
-                    HelpExampleCli("getmempooltxspendingprevout", "\"[{\\\"txid\\\":\\\"a08e6907dbbd3d809776dbfc5d82e371b764ed838b5655e72f463568df1aadf0\\\",\\\"vout\\\":3}]\"")
-                    + HelpExampleRpc("getmempooltxspendingprevout", "\"[{\\\"txid\\\":\\\"a08e6907dbbd3d809776dbfc5d82e371b764ed838b5655e72f463568df1aadf0\\\",\\\"vout\\\":3}]\"")
-                },
+            },
+        },
+        RPCResult{
+            RPCResult::Type::ARR, "", "",
+            {
+                {RPCResult::Type::OBJ, "", "",
+                {
+                    {RPCResult::Type::STR_HEX, "txid", "the transaction id of the spent output"},
+                    {RPCResult::Type::NUM, "vout", "the vout value of the spent output"},
+                    {RPCResult::Type::STR_HEX, "spendingtxid", /*optional=*/true, "the transaction id of the mempool transaction spending this output (omitted if unspent)"},
+                }},
+            }
+        },
+        RPCExamples{
+            HelpExampleCli("getmempooltxspendingprevout", "\"[{\\\"txid\\\":\\\"a08e6907dbbd3d809776dbfc5d82e371b764ed838b5655e72f463568df1aadf0\\\",\\\"vout\\\":3}]\"")
+            + HelpExampleRpc("getmempooltxspendingprevout", "\"[{\\\"txid\\\":\\\"a08e6907dbbd3d809776dbfc5d82e371b764ed838b5655e72f463568df1aadf0\\\",\\\"vout\\\":3}]\"")
+        },
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
-{
-    RPCTypeCheckArgument(request.params[0], UniValue::VARR);
-    const UniValue& output_params = request.params[0];
-    if (output_params.empty()) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, outputs are missing");
-    }
+        {
+            RPCTypeCheckArgument(request.params[0], UniValue::VARR);
+            const UniValue& output_params = request.params[0];
+            if (output_params.empty()) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, outputs are missing");
+            }
 
-    std::vector<COutPoint> outputs;
-    outputs.reserve(output_params.size());
+            std::vector<COutPoint> outputs;
+            outputs.reserve(output_params.size());
 
-    for (unsigned int idx = 0; idx < output_params.size(); idx++) {
-        const UniValue& o = output_params[idx].get_obj();
+            for (unsigned int idx = 0; idx < output_params.size(); idx++) {
+                const UniValue& o = output_params[idx].get_obj();
 
-        RPCTypeCheckObj(o,
-            {
-                {"txid", UniValueType(UniValue::VSTR)},
-                {"vout", UniValueType(UniValue::VNUM)},
-            });
+                RPCTypeCheckObj(o,
+                                {
+                                    {"txid", UniValueType(UniValue::VSTR)},
+                                    {"vout", UniValueType(UniValue::VNUM)},
+                                });
+
+                const uint256 txid(ParseHashO(o, "txid"));
+                const int nOutput = find_value(o, "vout").get_int();
+                if (nOutput < 0) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, vout cannot be negative");
+                }
 
-        const uint256 txid(ParseHashO(o, "txid"));
-        const int nOutput = find_value(o, "vout").get_int();
-        if (nOutput < 0) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, vout cannot be negative");
-        }
+                outputs.emplace_back(txid, nOutput);
+            }
 
-        outputs.emplace_back(txid, nOutput);
-    }
+            const CTxMemPool& mempool = EnsureAnyMemPool(request.context);
+            LOCK(mempool.cs);
 
-    const CTxMemPool& mempool = EnsureAnyMemPool(request.context);
-    LOCK(mempool.cs);
-
-    UniValue result{UniValue::VARR};
+            UniValue result{UniValue::VARR};
 
-    for (const COutPoint& output : outputs) {
-        UniValue o(UniValue::VOBJ);
-        o.pushKV("txid", output.hash.ToString());
-        o.pushKV("vout", (uint64_t)output.n);
+            for (const COutPoint& output : outputs) {
+                UniValue o(UniValue::VOBJ);
+                o.pushKV("txid", output.hash.ToString());
+                o.pushKV("vout", (uint64_t)output.n);
 
-        const CTransaction* spendingTx = mempool.GetConflictTx(output);
-        if (spendingTx != nullptr) {
-            o.pushKV("spendingtxid", spendingTx->GetHash().ToString());
-        }
+                const CTransaction* spendingTx = mempool.GetConflictTx(output);
+                if (spendingTx != nullptr) {
+                    o.pushKV("spendingtxid", spendingTx->GetHash().ToString());
+                }
 
-        result.push_back(o);
-    }
+                result.push_back(o);
+            }
 
-    return result;
-},
+            return result;
+        },
     };
 }
 
diff --git a/test/functional/rpc_mempool_info.py b/test/functional/rpc_mempool_info.py
index ef5b9535d4..da4314444d 100755
--- a/test/functional/rpc_mempool_info.py
+++ b/test/functional/rpc_mempool_info.py
@@ -13,19 +13,17 @@ from test_framework.util import (
     assert_raises_rpc_error,
     chain_transaction,
 )
+from test_framework.wallet import MiniWallet
+
 
 class RPCMempoolInfoTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 1
-        self.setup_clean_chain = True
-
-    def skip_test_if_missing_module(self):
-        self.skip_if_no_wallet()
 
     def run_test(self):
-        # Mine some blocks and have them mature.
-        self.generate(self.nodes[0], COINBASE_MATURITY + 1)
-        confirmed_utxo = self.nodes[0].listunspent(1)[0]
+        self.wallet = MiniWallet(self.nodes[0])
+        self.wallet.rescan_utxos()
+        confirmed_utxo = self.wallet.get_utxo()
 
         # Create a tree of unconfirmed transactions in the mempool:
         #             txA
@@ -41,15 +39,21 @@ class RPCMempoolInfoTest(BitcoinTestFramework):
         #            \   /
         #             \ /
         #             txH
-        fee = Decimal("0.0001")
-        (txidA, txA_amount) = chain_transaction(self.nodes[0], [confirmed_utxo['txid']], [0], confirmed_utxo['amount'], fee, 2)
-        (txidB, txB_amount) = chain_transaction(self.nodes[0], [txidA], [0], txA_amount, fee, 2)
-        (txidC, txC_amount) = chain_transaction(self.nodes[0], [txidA], [1], txA_amount, fee, 2)
-        txidD, _ = chain_transaction(self.nodes[0], [txidB], [0], txB_amount, fee, 1)
-        (txidE, txE_amount) = chain_transaction(self.nodes[0], [txidB], [1], txB_amount, fee, 3)
-        (txidF, txF_amount) = chain_transaction(self.nodes[0], [txidC], [0], txC_amount, fee, 2)
-        txidG, _ = chain_transaction(self.nodes[0], [txidC], [1], txC_amount, fee, 1)
-        txidH, _ = chain_transaction(self.nodes[0], [txidE, txidF], [0, 1], txE_amount + txF_amount, fee, 1)
+
+        def create_tx(**kwargs):
+            return self.wallet.send_self_transfer_multi(
+                from_node=self.nodes[0],
+                **kwargs,
+            )
+
+        txA = create_tx(utxos_to_spend=[confirmed_utxo], num_outputs=2)
+        txB, txC = [create_tx(utxos_to_spend=[u], num_outputs=2) for u in txA["new_utxos"]]
+        txD, txE = [create_tx(utxos_to_spend=[u]) for u in txB["new_utxos"]]
+        txF, txG = [create_tx(utxos_to_spend=[u]) for u in txC["new_utxos"]]
+        txH = create_tx(utxos_to_spend=txE["new_utxos"] + txF["new_utxos"])
+        txidA, txidB, txidC, txidD, txidE, txidF, txidG, txidH = [
+            tx["txid"] for tx in [txA, txB, txC, txD, txE, txF, txG, txH]
+        ]
 
         mempool = self.nodes[0].getrawmempool()
         assert_equal(len(mempool), 8)

MarcoFalke replied:
Good to see you fixed up the test. If you squash everything down into one commit, it should be easier for others to review. You don't need to mention me in the commit body.

t-bast replied:
Thanks, done!

vincenzopalazzo reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24408/commits/c3cde8ca864f13a638f558d1631f144ffa37e03d

t-bast commented:
@danielabrozzoni @w0xlt @glozow would you mind having a look at the latest state of this PR?

My only open question/thing to consider is whether I should include the whole spending tx instead of just the `spendingtxid` in the results. I chose to only include the `spendingtxid` because otherwise if a transaction appears multiple times (because it spends many of the provided prevouts) it would be duplicated, which is a bandwidth waste. Callers can chain this with `getmempoolentry` to get more complete tx data for the conflicting transaction. If you think it makes more sense to directly include the mempool tx let me know, I don't have very strong feelings about that.

vincenzopalazzo commented:
>My only open question/thing to consider is whether I should include the whole spending tx instead of just the spendingtxid in the results. I chose to only include the spendingtxid because otherwise if a transaction appears multiple times (because it spends many of the provided prevouts) it would be duplicated, which is a bandwidth waste. 

IMHO the RPC command needs to do one operation, and in a clean way. What `getmempooltxspendingprevout` already done with your actual implementation. How I can see in my mind if the `getmempooltxspendingprevout` return all the mem pool entry is a solution that makes the API a little bit messy?


A solution that can fix your open question is to add another RPC method that returns all the `spendingtxid` for each `prevouts` in there the duplicated can be fine because lives in a separate JSON object.

What do you think?

t-bast commented:
> IMHO the RPC command needs to do one operation, and in a clean way. What getmempooltxspendingprevout already done with your actual implementation. 

:+1: 

> A solution that can fix your open question is to add another RPC method that returns all the spendingtxid for each prevouts in there the duplicated can be fine because lives in a separate JSON object.

I would avoid adding yet another RPC method, since the same result can be easily achieved by doing `getmempooltxspendingprevout` followed by `getmempoolentry`. I would add new RPC methods only for things that really cannot be done with existing RPCs, otherwise `bitcoind` will end up with way too many RPC methods which is going to be a pain to maintain.

vincenzopalazzo commented:
>I would avoid adding yet another RPC method, since the same result can be easily achieved by doing getmempooltxspendingprevout followed by getmempoolentry. I would add new RPC methods only for things that really cannot be done with existing RPCs, otherwise bitcoind will end up with way too many RPC methods which is going to be a pain to maintain.

I have your same idea here!

glozow commented on test/functional/rpc_mempool_info.py:
```diff
@@ -0,0 +1,103 @@
+
+        self.log.info("Unknown input fields")
+        result = self.nodes[0].getmempooltxspendingprevout([ {'txid' : txidC, 'vout' : 1, 'unknown' : 42} ])
+        assert_equal(result, [ {'txid' : txidC, 'vout' : 1, 'spendingtxid' : txidG} ])
```
Question: would it be better for this to throw a `RPC_INVALID_PARAMETER`?

t-bast replied:
Good question. I tend to silently ignore unknown parameters in APIs because it's generally more friendly to clients. If a future version of `bitcoind` introduces new optional parameters, a client can safely start including them without caring about the version of `bitcoind` it's running against, and calls will work on older `bitcoind` as well. That's not a very strong reason though, I can change that if you want.

glozow replied:
No objections here! Being more user-friendly sounds good :grin:

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -587,6 +587,89 @@ static RPCHelpMan getmempoolentry()
+
+            UniValue result{UniValue::VARR};
+
+            for (const COutPoint& output : outputs) {
```
Small nit: since the type is `COutPoint`, I would prefer to name this `prevout` or `outpoint` so we don't confuse it with a `COutput`.

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/24408/commits/8c9152c1098acec4b1b4e333240f0e95eea21159

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -587,6 +587,89 @@ static RPCHelpMan getmempoolentry()
     };
 }
 
+static RPCHelpMan getmempooltxspendingprevout()
```
Also a nit (especially since that PR didn't exist when you opened this one): if we end up with #24539 as well, a more generic RPC named `gettxspendingprevout` might be better, i.e. to retrieve a transaction from block or from mempool.

t-bast replied:
Good idea! I changed that in https://github.com/bitcoin/bitcoin/pull/24408/commits/8c9152c1098acec4b1b4e333240f0e95eea21159 so that #24539 can build on top of it. I currently kept it in `mempool.cpp` and in the mempool RPCs list, I feel it's the responsibility of #24539 to move it elsewhere, is that ok @sstone?

I think #24539 won't have to modify any of the RPC fields, which is great. Callers will simply follow up with `getrawtransaction` which will provide the `confirmations` field, informing them of whether the sending transaction is in the mempool or is confirmed.

sstone replied:
Yes, and I agree that it makes sense to rebase https://github.com/bitcoin/bitcoin/pull/24539 on this PR

glozow commented on test/functional/rpc_mempool_info.py:
```diff
@@ -0,0 +1,103 @@
+        txB = create_tx(utxos_to_spend=[txA["new_utxos"][0]], num_outputs=2)
+        txC = create_tx(utxos_to_spend=[txA["new_utxos"][1]], num_outputs=2)
+        txD = create_tx(utxos_to_spend=[txB["new_utxos"][0]], num_outputs=1)
+        txE = create_tx(utxos_to_spend=[txB["new_utxos"][1]], num_outputs=3)
```
nit: seems like txE only needs 1 output

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/24408/commits/8c9152c1098acec4b1b4e333240f0e95eea21159

glozow reviewed (commented):
code review ACK c3cde8ca864f13a638f558d1631f144ffa37e03d. Looks correct, clean, and well-tested to me. Some nits if you retouch.

> My only open question/thing to consider is whether I should include the whole spending tx instead of just the spendingtxid in the results.

I don't feel strongly, but if you change your mind, I think it's perfectly fine to include the whole spending tx. There's not much waste since we've already retrieved it from the mempool, and only a privileged user can be using the RPC interface anyway.

t-bast committed:
Add RPC to get mempool txs spending outputs

We add an RPC to fetch the mempool transactions spending given outpoints.
Without this RPC, application developers would need to first call
`getrawmempool` which returns a long list of `txid`, then fetch each of
these txs individually to check whether they spend the given outpoint(s).

This RPC can later be enriched to also find confirmed transactions instead
of being restricted to mempool transactions.

t-bast commented:
> I don't feel strongly, but if you change your mind, I think it's perfectly fine to include the whole spending tx. There's not much waste since we've already retrieved it from the mempool, and only a privileged user can be using the RPC interface anyway.

I think it's better to only have the `spendingtxid`, as there is a lot of useful information in the output of `getrawtransaction` that callers will want anyway (so they will follow up with a call to that RPC), and we don't want to replicate all of it in the output of `gettxspendingprevout`. We're regularly seeing very big coinjoin transactions in our systems, so if we include the whole tx in the RPC response we can easily have responses that span several Mb, which is a waste for `bitcoind` nodes on remote machines accessed through a (sometimes flaky) VPN (which we've seen some lightning node operators do).

glozow commented:
re ACK 8c9152c1098acec4b1b4e333240f0e95eea21159

MarcoFalke commented:
cc Antoine for review (@darosior @ariard), if interested.

ariard commented on src/rpc/mempool.cpp:
```diff
@@ -587,6 +587,89 @@ static RPCHelpMan getmempoolentry()
+                {RPCResult::Type::OBJ, "", "",
+                {
+                    {RPCResult::Type::STR_HEX, "txid", "the transaction id of the spent output"},
+                    {RPCResult::Type::NUM, "vout", "the vout value of the spent output"},
```
super-nit: if you wanna normalize the wording, one of the information outcome of this RPC is that output might not be spent in fact, so might said "the check output" here and above.

t-bast replied:
Thanks, done in https://github.com/bitcoin/bitcoin/pull/24408/commits/8f9335f19ce75515f9e6944de503a5278a92f464

ariard reviewed (commented):
Code Review ACK 8c9152c

-----------

A caveat : the usage of this RPC you're describing sounds unsafe to me. If you're desirous to monitor the spend of a shared-utxo such as LN funding output, your counterparty might be able to broadcast multiple conflicting transactions. If this counterparty knows your full-node, your mempool can even be partitioned from the rest of the network, with a junk feerate, potentially wrecking your fee-bump. Of course, this assumes a bit of cross-layer mapping work on your malicious counterparty. So better to reconciliate from multiple mempools or opt for blind fee-bump after a while imho.

t-bast commented:
> A caveat : the usage of this RPC you're describing sounds unsafe to me. If you're desirous to monitor the spend of a shared-utxo such as LN funding output, your counterparty might be able to broadcast multiple conflicting transactions.

Thanks for the comments! But that shouldn't be an issue in our case (at least not related to this RPC). This RPC is helpful to react instantly once a shared output is spent. It doesn't matter if in the end it's another transaction that gets confirmed, we will learn this when we receive blocks and will then adapt. It then becomes a different problem of ensuring you're receiving blocks (avoiding complete eclipse attacks), which we solve by a combination of blockchain watchdogs and a good `bitcoind` setup.

Protecting against commit tx pinning is an entirely separate issue as well, which exists regardless of whether this RPC exists or not.

luke-jr commented:
>@luke-jr can you detail how that would work? 

You would add the txout you need to watch to the wallet (maybe we need to add a way to do that), then the wallet would scan every new block/tx internally, letting you know via listtransactions.

>it would be wasteful to need both a watch-only wallet and a regular wallet to make this use-case work?

At least legacy wallets support both normal and watch-only in the same wallet...

w0xlt commented on src/rpc/mempool.cpp:
```diff
@@ -587,6 +587,89 @@ static RPCHelpMan getmempoolentry()
+            for (unsigned int idx = 0; idx < output_params.size(); idx++) {
+                const UniValue& o = output_params[idx].get_obj();
+
+                RPCTypeCheckObj(o,
+                                {
+                                    {"txid", UniValueType(UniValue::VSTR)},
+                                    {"vout", UniValueType(UniValue::VNUM)},
+                                });
```
nit: With this change, RPC will not accept invalid/unrelated parameters.

```suggestion
                RPCTypeCheckObj(o,
                                /*typesExpected=*/{
                                    {"txid", UniValueType(UniValue::VSTR)},
                                    {"vout", UniValueType(UniValue::VNUM)},
                                }, /*fAllowNull=*/false, /*fStrict=*/true);
```

t-bast replied:
Thanks for the tip! I'm leaning more towards being flexible here with regards to unknown parameters (see [this comment](https://github.com/bitcoin/bitcoin/pull/24408#discussion_r857348954)) but it's a weak preference, if you feel more strongly about being strict here I can apply your proposed change, let me know!

w0xlt replied:
Both options work for me. I don't have a strong opinion on that.
I just assume that if an invalid parameter is used then maybe the user is expecting different behavior than the default.

t-bast replied:
Since both you and @glozow noticed this, it's probably worth making this stricter as you suggest. Done in https://github.com/bitcoin/bitcoin/pull/24408/commits/ad14372cab1897b3f5efafe9861ec696f5e97060

w0xlt commented on test/functional/rpc_mempool_info.py:
```diff
@@ -0,0 +1,103 @@
+        result = self.nodes[0].gettxspendingprevout([ {'txid' : txidB, 'vout' : 0}, {'txid' : txidG, 'vout' : 3} ])
+        assert_equal(result, [ {'txid' : txidB, 'vout' : 0, 'spendingtxid' : txidD}, {'txid' : txidG, 'vout' : 3} ])
+
+        self.log.info("Unknown input fields")
+        result = self.nodes[0].gettxspendingprevout([ {'txid' : txidC, 'vout' : 1, 'unknown' : 42} ])
+        assert_equal(result, [ {'txid' : txidC, 'vout' : 1, 'spendingtxid' : txidG} ])
```
```suggestion
        self.log.info("Unknown input fields")
        assert_raises_rpc_error(-3, "Unexpected key unknown", self.nodes[0].gettxspendingprevout, [{'txid' : txidC, 'vout' : 1, 'unknown' : 42}])
```

t-bast replied:
Thanks, done in https://github.com/bitcoin/bitcoin/pull/24408/commits/ad14372cab1897b3f5efafe9861ec696f5e97060

w0xlt reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/24408/commits/8f9335f19ce75515f9e6944de503a5278a92f464

I confirmed that the new RPC correctly returns the ids of transactions that spend the outpoints passed as a parameter.

t-bast commented:
@luke-jr thanks, I understand what you mean, I didn't know a wallet could be both normal and watch-only. This is something that could make sense, we are currently doing this scan outside of `bitcoind`, but having `bitcoind` do it for us could be useful.

Does it completely replace this PR though? I can see value in having both options (a dedicated RPC and a watch-only feature), the watch-only mode could be added afterwards (especially if we see people starting to rely on this new RPC).

darosior reviewed (commented):
Conceptually, this feature needs a transaction indexer and that's what a wallet is. I think ideally this feature should be part of the wallet as a more general "get spending tx by prevout" feature, rather than a very usecase-specific feature leveraging the mempool index.
However, it would be a bit convoluted to use at the moment. It would require them to import a new descriptor for each output they would need to claim, without (currently) the possibility to delete it. The number of descriptors under watch would therefore grow continuously as the software would be used.
Accounting for the immediate benefit to users, and the low cost of this RPC (it only calls into a single mempool function), i agree it can be merged. I believe longer term we should rather improve the wallet instead of providing (what i think are) specialized wallet functionalities at the node level.

light code review ACK 8f9335f19ce75515f9e6944de503a5278a92f464 -- the code makes sense to me and i didn't spot any issue. But i only quickly glanced at the diff.

*(For reference, we have a similar issue with `revaultd`: in order to fetch a spending transaction from the watchonly wallet whose txid is unknown we need to iterate on the list of the wallet transactions. https://github.com/revault/revaultd/blob/b310208a8b3b420f9b0cddcc03588e264a99538b/src/bitcoind/interface.rs#L777-L864 (fortunately `listsinceblock` allows us to reduce the cost by only iterating on transactions since the last pool).)*

t-bast commented:
> Conceptually, this feature needs a transaction indexer and that's what a wallet is. I think ideally this feature should be part of the wallet as a more general "get spending tx by prevout" feature, rather than a very usecase-specific feature leveraging the mempool index.

I agree, and you're one step ahead, see #24539 :)
Thanks for the comments!

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/24408/commits/ad14372cab1897b3f5efafe9861ec696f5e97060

t-bast commented:
Simple rebase on `master` without any changes in https://github.com/bitcoin/bitcoin/pull/24408/commits/418557034055f740951294e7677ae9fd5149ea9b

t-bast commented:
I don't have any pending issues on that PR, if @ariard @darosior @glozow @MarcoFalke you have time to give this another pass whenever possible it would be great :pray:

darosior reviewed (commented):
re-utACK 418557034055f740951294e7677ae9fd5149ea9b

vincenzopalazzo reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/pull/24408/commits/418557034055f740951294e7677ae9fd5149ea9b

danielabrozzoni reviewed (approved):
re-tACK 418557034055f740951294e7677ae9fd5149ea9b

fanquake commented:
cc @glozow

w0xlt reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/24408/commits/418557034055f740951294e7677ae9fd5149ea9b
